Enclave over WebSockets

melowe opened the pull request:
Currently dormant module enabling the remoting of enclave over WebSockets in the same way as enclave-jaxws remotes over rest/http. EnclaveEndpointTest fires up local WebSockets server and delegates to mocked enclave instance.

melowe committed:
Inject enclave cli adapter for standalone enclave instance.
Move all enclave related config out of core into enclave api module . 
Add tactical workaround for handing null always sendto list in the form 
of a helper class.

melowe committed:
Add jaxb provider  property for all integration tests. Add
integration test for enclave standalone.

melowe committed:
Create initial module for refactoring peer to peer party info 
communication to be stateful.

melowe committed:
Initial module to provide access to enclave as websocket service.

melowe committed:
Add builder for cleaner object construction and encapsulate Nonce object.

melowe committed:
Add arg type enum to decrible args in enclave functions. 
Encode bytearrays to base64 and provide nessesary codecs. 
Use json as transport format rarher than object serialisation.

melowe committed:
Decorate request type in adapter rather than in each object. Use 
codec class name as discriminator value as all values being decoded
are encoded by the same object.

melowe committed:
Rename request type to encodedBy and use Simple Class name was value

melowe committed:
More tests

melowe committed:
Yet more tests and logging tweaks. 
Leave default error handing for server errors until we require anything 
else.

melowe committed:
Yet more tests and logging tweaks.

melowe committed:
Typo

melowe committed:
Add builder for raw transaction

melowe committed:
Added encrypt raw transaction payload.

melowe committed:
Add encrypt raw function

melowe committed:
Add unencryptTransaction  function

melowe committed:
Add create recipient box  function

melowe committed:
add tests for last coverage issues.

melowe committed:
Merge branch 'master' into feature/enclave

prd-fox commented on enclave/enclave-api/src/main/java/com/quorum/tessera/enclave/AlwaysSendToHelper.java:
```diff
@@ -0,0 +1,27 @@
+package com.quorum.tessera.enclave;
```
What issue is this really addressing? What was wrong about the configuration passing in the list directly? 

As far as I can see, the configuration object is still in scope for both the main application and the standalone enclave, so it's not a scoping issue. This also just sits in-between the config and the key converter, so it doesn't seem to be a circular reference issue either.

prd-fox commented on enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/AlwaysSendToHelperTest.java:
```diff
@@ -0,0 +1,20 @@
+package com.quorum.tessera.enclave;
```
Pending above issue, this could be removed.

prd-fox commented on enclave/enclave-websockets/src/main/java/com/jpmorgan/quorum/enclave/websockets/EnclaveEndpoint.java:
```diff
@@ -0,0 +1,149 @@
+
+        final WebSocketTemplate webSocketTemplate = new WebSocketTemplate(session);
+
+        if (type == EnclaveRequestType.DEFAULT_PUBLIC_KEY) {
```
Are `if` statements really the best option here? A `switch` statement I think it the better alternative, either directly or passing off to private methods.
This method is doing too many things otherwise.

prd-fox commented on enclave/enclave-websockets/src/main/java/com/jpmorgan/quorum/enclave/websockets/EnclaveEndpoint.java:
```diff
@@ -0,0 +1,149 @@
+            });
+            return;
+        }
+        throw new UnsupportedOperationException(type + " is not supported");
```
Can we send back an error to the caller?

prd-fox commented on enclave/enclave-websockets/src/main/java/com/jpmorgan/quorum/enclave/websockets/EnclaveRequestCodec.java:
```diff
@@ -0,0 +1,136 @@
+
+            ArgType type = enclaveRequestType.getParamTypes().get(i);
+
+            if (type == ArgType.PUBLIC_KEY_LIST) {
```
A `switch` would be better here.

prd-fox committed:
Force the "alwaysSendTo" list to be initialised in order to remove the
helper object.

Use switch statements instead of multiple if/elses for enum values.

prd-fox committed:
Merge branch 'master' into feature/enclave

prd-fox committed:
Fix checkstyle issue

Krish1979 committed:
Merge branch 'master' into feature/enclave
